File: frame/src/main/java/com/guoliang/frame/base/ActivityManager.java
package com.guoliang.frame.base;

import android.app.Activity;
import android.os.Process;

import java.lang.ref.WeakReference;
import java.util.ListIterator;
import java.util.Stack;

/**
 * @Description:
 * @Author: zhangguoliang
 * @CreateTime: 2020/5/26 17:14
 */
public class ActivityManager {

    private ActivityManager() {
    }

    /**
     * ActivityManager实例
     */
    private static ActivityManager INSTANCE;

    public static ActivityManager getInstance(){
        if (INSTANCE==null){
            synchronized (ActivityManager.class){
                INSTANCE = new ActivityManager();
            }
        }
        return INSTANCE;
    }


    Stack<WeakReference<Activity>> stack=null;

    /**
     * 获取栈顶Activity（堆栈中最后一个压入的）
     *
     * @return Activity
     */
    public Activity getTopActivity(){
        if (stack!=null){
            return stack.lastElement().get();
        }else {
            return null;
        }
    }

    /**
     * 添加Activity到堆栈
     */
    public void addActivity(Activity activity){
        if (stack==null){
            stack=new Stack<>();
        }
        stack.add(new WeakReference<>(activity));
    }

    /**
     * 删除Activity
     *
     * @param activity 指定的Activity
     */
    public void removeActivity(Activity activity){
        if (stack!=null){
            stack.remove(new WeakReference<>(activity));
        }
    }

    /**
     * 通过class 获取栈顶Activity
     *
     * @param aClass Activity的class
     * @return Activity
     */
    public Activity getActivityByClass(Class aClass){
        for (WeakReference<Activity> activityWeakReference : stack) {
            if (activityWeakReference.get().getClass() == aClass){
                return activityWeakReference.get();
            }
        }
        return null;
    }

    /**
     * 结束栈顶Activity（堆栈中最后一个压入的）
     */
    public void killTopActivity(){
        if (stack!=null){
            killActivity(stack.lastElement().get());
        }
    }

    /**
     * 结束指定的Activity
     *
     * @param activity Activity
     */
    public void killActivity(Activity activity){
        if (stack!=null) {
            for (WeakReference<Activity> reference : stack) {
                if (reference.get() == activity) {
                    stack.remove(reference);
                    if (activity!=null&&!reference.get().isFinishing()) {
                        reference.get().finish();
                    }
                }
            }
        }
    }

    /**
     * 结束指定类名的Activity
     *
     * @param aClass Class
     */
    public void killActivity(Class aClass){
        if (stack!=null){
            for (WeakReference<Activity> reference : stack) {
                if (reference.get().getClass() == aClass) {
                    stack.remove(reference);
                    if (!reference.get().isFinishing()) {
                        reference.get().finish();
                    }
                }
            }
        }
    }

    /**
     * 结束所有Activity
     */
    public void killAllActivity(){
        if (stack!=null){
            ListIterator<WeakReference<Activity>> listIterator = stack.listIterator();
            while (listIterator.hasNext()){
                Activity activity = listIterator.next().get();
                if (activity!=null&&!activity.isFinishing()){
                    activity.finish();
                    listIterator.remove();
                }
            }
        }
    }

    /**
     * 移除除了某个activity的其他所有activity
     *
     * @param aClass Class
     */
    public void killAllActivityExceptOne(Class aClass){
        if (stack!=null) {
            ListIterator<WeakReference<Activity>> listIterator = stack.listIterator();
            while (listIterator.hasNext()){
                if (listIterator.next().get().getClass()==aClass){
                    continue;
                }else {
                    killActivity(listIterator.next().get());
                    listIterator.remove();
                }
            }
        }
    }

    /**
     * 退出应用程序
     */
    public void appExit(){
        killAllActivity();
        Process.killProcess(Process.myPid());
    }

    /**
     * 栈中Activity的数
     *
     * @return Activity的数
     */
    public int getStackActivitySize(){
        if (stack!=null){
            return stack.size();
        }else {
            return 0;
        }
    }


}

File: frameKT/src/main/java/com/guoliang/framekt/base/ActivityManager.kt
package com.guoliang.framekt.base

import android.app.Activity
import android.os.Process
import java.lang.ref.WeakReference
import java.util.*

/**
 * @Description:
 * @Author: zhangguoliang
 * @CreateTime: 2020/5/26 17:14
 */
class ActivityManager private constructor() {
    private var stack: Stack<Activity> = Stack()

    /**
     * 获取栈顶Activity（堆栈中最后一个压入的）
     *
     * @return Activity
     */
    val topActivity: Activity
        get() = stack.lastElement()

    /**
     * 添加Activity到堆栈
     */
    fun addActivity(activity: Activity) {
        stack.add(activity)
    }

    /**
     * 删除Activity
     *
     * @param activity 指定的Activity
     */
    fun removeActivity(activity: Activity) {
        stack.remove(activity)
    }

    /**
     * 通过class 获取栈顶Activity
     *
     * @param aClass Activity的class
     * @return Activity
     */
    fun getActivityByClass(aClass: Class<*>): Activity? {
        for (activityWeakReference in stack) {
            if (activityWeakReference.javaClass == aClass) {
                return activityWeakReference
            }
        }
        return null
    }

    /**
     * 结束栈顶Activity（堆栈中最后一个压入的）
     */
    fun killTopActivity() {
        killActivity(stack.lastElement())
    }

    /**
     * 结束指定的Activity
     *
     * @param activity Activity
     */
    fun killActivity(activity: Activity?) {
        for (reference in stack) {
            if (reference === activity) {
                stack.remove(reference)
                if (activity != null && !reference!!.isFinishing) {
                    reference.finish()
                }
            }
        }
    }

    /**
     * 结束指定类名的Activity
     *
     * @param aClass Class
     */
    fun killActivity(aClass: Class<*>) {
        for (reference in stack) {
            if (reference!!.javaClass == aClass) {
                stack.remove(reference)
                if (!reference.isFinishing) {
                    reference.finish()
                }
            }
        }
    }

    /**
     * 结束所有Activity
     */
    fun killAllActivity() {
        val listIterator = stack.listIterator()
        while (listIterator.hasNext()) {
            val activity = listIterator.next()
            if (activity != null && !activity.isFinishing) {
                activity.finish()
            }
        }
        stack.removeAllElements()
    }

    /**
     * 移除除了某个activity的其他所有activity
     *
     * @param aClass Class
     */
    fun killAllActivityExceptOne(aClass: Class<*>) {
        val listIterator = stack.listIterator()
        while (listIterator.hasNext()) {
            if (listIterator.next().javaClass == aClass) {
                continue
            } else {
                killActivity(listIterator.next())
                listIterator.remove()
            }
        }
    }

    /**
     * 退出应用程序
     */
    fun appExit() {
        killAllActivity()
        Process.killProcess(Process.myPid())
    }

    /**
     * 栈中Activity的数
     *
     * @return Activity的数
     */
    val stackActivitySize: Int
        get() = stack.size

    companion object {
        val instance = ActivityManagerHolder.holder
    }
    private object ActivityManagerHolder {
        val holder = ActivityManager()
    }
}